jobs/deploy-webapp.groovy

```diff
@@ -519,13 +542,75 @@ def _manualPromptCheck(prompt){
+// to trigger the jobs to abort. Let the caller decide if this exception
+// should bubble up the stack or not.
+def _maybeAbortJob(oldBuildResult, newBuildResult, reason) {
+   // NOTE(miguel): Build status changed, let's check if the job was
+   // aborted.  We should not be in an aborted state here but we have a
+   // theory that the job was aborted but jenkins did not successfully
+   // abort the thread because it was in the middle of uninterruptible
+   // work like making an HTTP request. If that's the case we are going
+   // to cause the job to abort from here to prevent new deploy-webapp
+   // jobs from getting stuck.
+   if (oldBuildResult != newBuildResult) {
+      // Because this is a theory that build are getting stuck because the
+      // signal to abort the current thread while we are in an uninterruptible
+      // cycle, we are not 100% sure what the correct things to check are
+      // to make an exact desicion. So we are going to log a message to slack
+      // to keep track of these cases.
+      _simpleSlackAlert(
+         "#infrastructure-deploys",
+         """Deploy aborted but thread is still running. We most likely have a 
+job that is stuck. ${reason}. prev state ${oldBuildResult}. current state 
+${newBuildResult}. ${env.BUILD_URL}. 
+See https://khanacademy.atlassian.net/wiki/spaces/INFRA/pages/2470543393/Stuck+jenkins+deploy-webapp+builds 
+for details to get information about stuck builds. cc @deploy-support""",
+         "error",
+      )
+
+      if (newBuildResult == "ABORTED") {
+         // Jenkins would throw a hudson.AbortException.  But to make
+         // things more clear that the job is aborting itself we will
+         // throw a different exception.  This exception will (should)
+         // be handle by the caller to ensure we clean things up.
+         // TODO(miguel): enable once we are confident this will work.
+         // Let's run it without auto aborting to get some ideas so that
+         // we don't mistakenly start aborting false positives.
+         // throw new AbortDeployJob("Deploy was aborted. " + reason)
+      }
+   }
+}
+
 // TODO(jacqueline): Make this a shared func with verifying smoke tests.
 def verifyPromptConfirmed(prompt, buildmasterFailures=0) {
+   if (currentBuild.result == "ABORTED") {
+      // TODO(miguel): perhaps throw exception once we understand better why
+      // we got into this function with an aborted job.
+      echo("Job is aborted but we are still trying to verify prompt status "+
+           "${prompt}. That's not great at all. Ping @deploy-support to "+
+           "take a look.")
+   }
+
    def status;
    while (status != "confirmed") {
+      def buildResult = currentBuild.result
       status = buildmaster.pingForPromptStatus(prompt,
                                                params.GIT_REVISION)
 
+      _maybeAbortJob(
+         buildResult,
+         currentBuild.result,
```

Comment:
I think it's pretty unlikely the job state will change will prompting for status. The change is most likely to occur because of a buildmaster or manual abort, which can happen at any time. If the current build is aborted, is the result really "ABORTED"? If so, that's great and it seems like we could just check if the current build has that (like is done in the `if` at the top of the function).

Reply:
Yes i just left a comment around that in this bit of code.  Jenkins logs do show that even tho the job is stuck, jenkins did try to abort the job.  At some point i was able to reproduce the issue and the call to `api/json` reported an aborted status even tho the build was still looping.  So unless the API is returning something that isn't matching the actual state of jenkins, this should work.   But im going to setup a way to test this more directly.